### PR TITLE
Attention vis

### LINKED ISSUE #42: Add Analysis for Comparing Adapter Attention Weights

### PR DESCRIPTION
**Pull request by sebastian9991**
Some visualizations for attention weights.

### CODE REVIEW

**Comment by sebastian9991** [comment]
Don't quite understand the difference between pytorch_adapter.bin & pytorch_model_head.bin. Used the latter for weights.

**Comment by sebastian9991** [comment]
Can add other visualization methods. e.g pca or compare cosine sim between paired adapters

File: figs/adapter_attention_weights/charts_attention_weights/opacity_pred_attention_weights.png
Comment (Jacob-Chmura) [inline figs/adapter_attention_weights/charts_attention_weights/opacity_pred_attention_weights.png:1]: It's not clear how I am to interpret these charts, as you mentioned, some sort of dim reduction or smoothing combined with an explicit way to compare **between** adapters is the important part

Comment (Jacob-Chmura) [reply figs/adapter_attention_weights/charts_attention_weights/opacity_pred_attention_weights.png:1]: what are these axes representing?

File: figs/adapter_attention_weights/charts_attention_weights/weight_pred_attention_weights.png
Comment (Jacob-Chmura) [inline figs/adapter_attention_weights/charts_attention_weights/weight_pred_attention_weights.png:1]: Why does this one look so different?

File: figs/adapter_attention_weights/hist_combined_weights/mobility_pred_combined_weights.png
Comment (Jacob-Chmura) [inline figs/adapter_attention_weights/hist_combined_weights/mobility_pred_combined_weights.png:1]: In isolation these are not very useful. We need to compare between adapters, and try to see if we can highlight general distributional patterns

File: scripts/analyze_attention_weights.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+parser.add_argument(
+    '--config-file',
+    type=str,
+    default='config/base.yaml',
+    help='Path to yaml configuration file for iterating tasks.',
+)
+parser.add_argument(
+    '--weights-dir',
+    type=str,
+    default='../weights/',
```
Comment (Jacob-Chmura) [inline scripts/analyze_attention_weights.py]: Can you please mimic my other scripts and make all these relative to the global root directory, and the resolve the paths as I have done, i.e.: https://github.com/Jacob-Chmura/plausibility-vaccine/blob/c1ac47817b351633a888545abb0b01b55582a2a5/scripts/generate_selectional_association_data.py#L46

**Review by Jacob-Chmura — COMMENTED** [review]
At the moment, these charts are not clear enough to add into the report, but it's a decent start. Have you looked at packages that do this for you? I just quickly found this one: https://github.com/jessevig/bertviz

The distribution plots are probably easier to get right, but they need to be contrasted across adapters. It may also be the case that to see any signal we need to condition on specific s-v-o samples to visualize are adapters firing. We need to think more about this

File: figs/adapter_attention_weights/hist_combined_weights/mobility_pred_combined_weights.png
Comment (sebastian9991) [reply figs/adapter_attention_weights/hist_combined_weights/mobility_pred_combined_weights.png:1]: Right. I'll try some other vis method to compare it.

**Review by Jacob-Chmura — CHANGES_REQUESTED** [review]
See me above comments for changes

**Comment by sebastian9991** [comment]
I was able to use bertviz. There is a single HTML output you can checkout in the /figs folder. I'll try what I mention about looking at attention on plausible/implausible sentences. I'll also look at ways to aggregate the results for better viewing. Bertviz seems like what would be best to use. I'll investigate it more. Pushing so you can see how it looks so far. You'll have to view the HTML in a browser I believe. 

I also tried a PCA method on a single adapter weight (not comparing them) It didn't reveal much to me.  Will explore ways to compare PCA


Note I skipped the trailing-whitespaces hook as it complained about the HTML file.

File: scripts/analyze_attention_weights.py (hunk 1 above)
Comment (sebastian9991) [reply scripts/analyze_attention_weights.py]: Updated it. Let me know if that's what you were looking for.

**Review by Jacob-Chmura — COMMENTED** [review]
Given our 5 page limit, and the relative importance of the other charts, I'm not sure this will end up going in the report unless we find a very interesting result. Otherwise, I think just the mispredictoins analysis will be sufficient for the discussion section.

**Pull request closed without merging**